backend/controllers/authController.js

```diff
@@ -30,6 +41,63 @@ const authController = {
       });
     }
   },
+
+  async verifyUser(req, res, next) {
+    const { username, password } = req.body[0];
+    let hashedPassword;
+    try {
+      if (!password) throw new Error('empty password');
+      await User.findOne({ username }, (err, userAccount) => {
+        if (err || !userAccount) {
+          res.redirect('/signup');
+        } else {
+          hashedPassword = userAccount.password;
+          bcrypt.compare(password, hashedPassword, function (err, result) {
+            // result == true
+            if (result === true) {
+              //after verification, pass user information to the next middleware
+              const user = {
+                username: userAccount.username,
+                id: userAccount._id.toString(),
+                gallery: userAccount.gallery,
+              };
+              res.locals.user = user;
+              console.log(user);
+              return next();
+            } else {
+              console.log('wrong password');
+              res.send('wrong password');
+            }
+          });
+        }
+      }).clone();
```

Comment:
What does the clone method do?

Reply:
We were getting the error below and used clone to work around it. Mongoose doesn't allow executing the same query object twice. We didn't see this ran elsewhere though so we chained clone

MongooseError: Query was already executed: User.findOne({somedata})